Add best practices page (#6478)
jenkins-infra/jenkins.io: content/doc/book/using/best-practices.adoc

This section describes best practices to provide executives, business managers, software developers, and architects with a better sense of what Jenkins can contribute throughout the project lifecycle.

== Use organization folders

Organization Folders offer a convenient way to allow Jenkins to automatically create and delete jobs as repositories and branches are created and deleted.
Particularly, if you use GitHub organizations, Bitbucket teams, GitLab groups, or Gitea organizations, then Jenkins can automatically detect the creation of a new repository and create a Multibranch Pipeline project for it.

See the link:/doc/book/pipeline-as-code/[organization folders] documentation for more details.

.GitHub organization folders
video::LbXKUKQ24T8[youtube,width=800,height=420]

.GitLab group folders
video::it6TOeQ6EHg[youtube,width=800,height=420]

.Bitbucket project folders
video::85b6fiVolfk[youtube,width=800,height=420]

.Gitea organization folders
video::NO3sZWRxgQM[youtube,width=800,height=420]

== Use multibranch Pipelines

If something prevents you from using organization folders, then use multibranch Pipelines.

> Insert more wisdom here ...

Author: This is a reminder that more text is needed here.  I would like that additional text before we merge.

> Insert more wisdom here ...

Author: Wisdom has now been inserted there.

See the link:/doc/book/pipeline-as-code/[multibranch Pipelines] documentation for more details.

== Use Pipeline

Insert Pipeline wisdom here

See the link:/doc/book/pipeline/[Pipeline] documentation for more details.

video::IOUm1lw7F58[youtube,width=800,height=420]

== Use reports effectively

Charts and graphs show trends and patterns that can highlight project status and progress.
Automated test results for unit tests, integration tests, and end to end tests can highlight brittleness or instability.
Coverage reports can illustrate areas that automated tests are not executing.
Compiler warning messages can be a first indication of a problem.
Static analysis tools can report risky code or code that may have security risks.
Performance test results can idenitfy delays or problem areas.

> The plugin:warnings-ng[warnings next generation] plugin provides convenient access to many reports, including:
> 
> * compiler warnnings and errors (like gcc, clang, javac, golang, etc.)
> * static analysis warnings and errors (like spotbugs, checkstyle, pmd, lint, cpd, Simian, etc.)
> * code coverage reports
> 
> video::tj3xYFA6Q2o[youtube,width=800,height=420]

Reviewer: ```suggestion
The plugin:warnings-ng[Warnings Next Generation] plugin provides convenient access to many reports, including:

* Compiler warnnings and errors (like gcc, clang, javac, or  golang)
* Static analysis warnings and errors (spotbugs, checkstyle, pmd, lint, cpd, or Simian)
* Code coverage reports

.How to use the Warnings Next Generation plugin 
video::tj3xYFA6Q2o[youtube,width=800,height=420]
```
suggestion to use the capitalized name for the plugin, and adjusting the text for some standardization/capitalization

== Show failures to the right people

> Tune the notifications from failing and unstable jobs so that the notifications reach the right people without distracting others.
> Many Jenkins users prefer to only be notified when a failure is likely their fault.
> They correctly assume that if they did not cause the failure, they are not the best person to investigate the failure.
> 
> Refine your notification system so that new test failures first notify the most recent committers, since they are the most likely cause of the issue.
> 
> .Sending Slack notifications
> video::EDVZli8GdUM[youtube,width=800,height=420]

Reviewer: ```suggestion
Configure the notifications from failing and unstable jobs so the notifications reach the right people without distracting others.
Many Jenkins users prefer to only be notified when a failure is likely their fault.
They correctly assume that if they did not cause the failure, they are not the best person to investigate the failure.

Refine your notification system so that new test failures first notify the most recent committers, since they are the most likely cause of the issue.

.Sending Slack notifications
video::EDVZli8GdUM[youtube,width=800,height=420]
```
suggestion to update some of the syntax

== Build on agents

Use agents to perform builds instead of running builds on the controller.

> It is much safer to use agents and much more scalable.

Reviewer: ```suggestion
It is much safer, and more scalable, to use agents.
```
suggestion to reformat sentence/syntax

More details are available in the link:/doc/book/security/controller-isolation/[controller isolation] documentation.

== Back up regularly

Even the most reliable of systems will have failures.
Prepare and regularly check the health of your backups.

> More details in the link:/doc/book/system-administration/backing-up/[backup documentation].

Reviewer: ```suggestion
More details can be found in the link:/doc/book/system-administration/backing-up/[backup documentation].
```
suggestion to make sentence flow better

> == Use simple project names

Reviewer: ```suggestion
== Simple project names
```
similar to previous suggestion for section titles

Jenkins uses project names for folders related to the project.
Many tools cannot handle spaces, dollar signs, or similar characters in file paths.

> Limit yourself to alphanumeric characters [a-zA-Z0-9_-]+ in project names.

Reviewer: ```suggestion
Limit yourself to alphanumeric characters [a - z, A - Z, 0 - 9, _, -, +] in project names.
```
suggestion to reformat characters for clarity

Author: The string is formatted as though it were a regular expression.  I'd rather remove the regex syntax than insert the white space, since that makes it no longer a regex.

Reviewer: That makes sense to me @MarkEWaite! My original intent was to ensure that the `+` was contained within the brackets, so I should have just left it at that.

> Use the Display Name feature to make them look nice.

Reviewer: ```suggestion
Use the *Display Name* feature to make them look nice.
```
suggestion to bold UI

The "Restrict project naming" setting in the system configuration can enforce naming restrictions for all projects.

== Fingerprint your dependencies

When you have interdependent projects, it often becomes hard to keep track of which version of this is used by which version of that.

> Jenkins supports "file fingerprinting" to simplify this, so make the best use of it.
> 
> More information is available in the link:/doc/book/using/fingerprints/[fingerprinting page].

Reviewer: ```suggestion
Jenkins supports "file fingerprinting" to simplify this.

More information is available on the link:/doc/book/using/fingerprints/[fingerprinting page].
```
suggestion to update some text/remove unnecessary text

== Strive for reproducible builds

Reproducible builds are a set of software development practices that create an independently-verifiable path from source to binary code.

> A continuous integration build is a possible first step towards reproducible builds, but is usually not sufficient

Reviewer: ```suggestion
A continuous integration build is a potential first step towards reproducible builds, but is usually not sufficient.
```
suggestion to update punctuation/syntax

> See the link:https://reproducible-builds.org/[reproducible builds site] for more details.

Reviewer: ```suggestion
Refer to the link:https://reproducible-builds.org/[reproducible builds site] for more details.
```
same as previous suggestions to use"Refer" instead of "See"

> // == Archive unused jobs
> //
> // Unused jobs should be archived, so they can be restored if the need arises.
> //
> // See link:/doc/book/system-administration/administering-jenkins[Administering Jenkins] for ways to do this.

Reviewer: ```suggestion
```
would it be worth removing this if it is commented out? If there is other documentation that would work in its place, then it should be updated to reflect that.

== Prevent resource collisions in parallel jobs

> Multiple jobs running at the same time often cause collisions if they set up some kind of service, or need exclusive access.

Reviewer: ```suggestion
Multiple jobs running at the same time often cause collisions if they set up a service or need exclusive access.
```
small suggestion for syntax

If your builds involve the use of databases or other networked services, you need to ensure that they don't interfere with each other.
Allocate a different port for parallel project builds to avoid build collisions.

> If that's not possible (e.g. in the case of a persistent resource that needs to be locked) you can prevent builds that use it from running at the same time using e.g. link:https://github.com/jenkinsci/throttle-concurrent-builds-plugin[Throttle Concurrent Builds Plugin.]

Reviewer: ```suggestion
If that's not possible, for example in the case of a persistent resource that needs to be locked, you can prevent builds that use it from running at the same time using the link:https://github.com/jenkinsci/throttle-concurrent-builds-plugin[Throttle Concurrent Builds] plugin.
```
Suggestion to update syntax

== Avoid scheduling overload

Schedule jobs to balance the number of jobs running concurrently.
If you're using timer triggers or are periodically polling, use the `H` syntax in the cron expression to introduce scheduing jitter.

> The predefined tokens such as @hourly can also be used to distribute job starting times more evenly.

Reviewer: ```suggestion
The predefined tokens such as `@hourly` can also be used to distribute job starting times more evenly.
```
suggestion to make token more obvious using code formatting

---
layout: section
title: Best Practices
---
ifdef::backend-html5[]
:description:
:author:
:sectanchors:
:toc:
:toclevels: 4
:hide-uri-scheme:
ifdef::env-github[:imagesdir: ../resources]
ifndef::env-github[:imagesdir: ../../resources]
endif::[]

Continuous Integration (CI) with automated test execution and trends has changed the way companies look at build management, release management, deployment automation, and test orchestration.
This section describes best practices to provide executives, business managers, software developers, and architects with a better sense of what Jenkins can contribute throughout the project lifecycle.

== Automate job definition

Jenkins can automatically create, update, and delete jobs based on the repositories it detects in your software configuration management system.

> Structure your job definitions to get the most benefit from automatic job management by Jenkins.

Reviewer: ```suggestion
Leverage this feature and optimize your job management by structuring your job definitions in a way that maximizes the benefits offered by Jenkins' automatic job management capabilities.
```
alternative suggestion, the idea is to make it more action oriented/putting it on the user that they can/should do this

There are multiple alternatives for automatic job management, including:

> * <<Use organization folders>> - create, update, and delete multibranch Pipeline folders and Pipeline jobs automatically (preferred)

Author: The <<Use organization folders>> markup is required as a link to the later heading in the document.  I liked it because I thought it was more action oriented.  I think that a best practices page is the right place to be more explicit on actions that the reader should take.

> * <<Use multibranch Pipelines>> - create, update and deleted Pipeline jobs automatically

Author: I prefer to not restate the left hand text ("multibranch Pipelines") in the right hand side.

* <<Use Pipeline>> - manually defined Pipeline jobs

=== Use organization folders

Organization Folders offer a convenient way to allow Jenkins to automatically create and delete jobs as repositories and branches are created and deleted.
If you use GitHub organizations, Bitbucket teams, GitLab groups, or Gitea organizations, then Jenkins can automatically detect the creation of a new repository and create a Multibranch Pipeline project for it.

Refer to the link:/doc/book/pipeline/pipeline-as-code/#organization-folders[organization folders] documentation for more details.

.GitHub organization folders
video::LbXKUKQ24T8[youtube,width=800,height=420]

.GitLab group folders
video::it6TOeQ6EHg[youtube,width=800,height=420]

.Bitbucket project folders
video::85b6fiVolfk[youtube,width=800,height=420]

.Gitea organization folders
video::NO3sZWRxgQM[youtube,width=800,height=420]

=== Use multibranch Pipelines

> If something prevents you from using organization folders, then use multibranch Pipelines.

Reviewer 2: ```suggestion
If you are unable to use organization folders, you can opt for multibranch Pipelines as an alternative.
```

> Organization folders are preferred over multibranch Pipelines because they will automatically create and delete multibranch projects when repositories are created or deleted.

Reviewer 2: ```suggestion
However, it's important to note that organization folders are preferred over multibranch Pipelines because they provide the automation of creating and deleting multibranch projects when repositories are added or removed.
```

.GitHub multibranch Pipelines
video::aDmeeVDrp0o[youtube,width=800,height=420]

.GitLab group folders
video::y4XGFluzPHY[youtube,width=800,height=420]

.Bitbucket project folders
video::LNfthmZuRDI[youtube,width=800,height=420]

Refer to the link:/doc/book/pipeline/pipeline-as-code/#multibranch-pipeline-projects[multibranch Pipelines] documentation for more details.

=== Use Pipeline

> If something prevents you from using organization folders and multibranch Pipelines, use Pipeline jobs.

Reviewer 2: ```suggestion
If organization folders are not an option for you, you can consider using multibranch Pipelines as an alternative.
```

> Organization folders and multibranch Pipelines are preferred over Pipeline jobs because they will automatically create and delete multibranch projects when repositories are created or deleted.

Reviewer 2: ```suggestion
However, it's important to highlight that organization folders are preferred due to their ability to automatically create and delete multibranch projects when repositories are added or removed.
```

Refer to the link:/doc/book/pipeline/[Pipeline] documentation for more details.

.Differences between Freestyle and Pipeline in Jenkins
video::IOUm1lw7F58[youtube,width=800,height=420]

== Manage your jobs

> Jenkins job definitions can be managed and improved to make user interactions smoother and more productive.

Reviewer 2: ```suggestion
Jenkins job definitions can be managed and optimized to enhance user interactions and productivity.
```

=== Report build results

> Charts and graphs show trends and patterns that can highlight project status and progress.

Reviewer 2: ```suggestion
Charts and graphs provide valuable insights into project status and progress, showcasing trends and patterns.
```

> Automated test results for unit tests, integration tests, and end-to-end tests can highlight brittleness or instability.

Reviewer 2: ```suggestion
Automated test results, including unit tests, integration tests, and end-to-end tests, can reveal brittleness or instability. 
```

> Coverage reports can illustrate areas where automated tests are not executing.

Reviewer 2: ```suggestion
Coverage reports help identify areas where automated tests are not being executed.
```

> Compiler warning messages can be the first indication of a problem.

Reviewer 2: ```suggestion
Compiler warning messages often serve as the first indication of a problem.
```

> Static analysis tools can report risky code or code that may have security risks.

Reviewer 2: ```suggestion
Static analysis tools are effective in reporting risky code or code with potential security risks.
```

> Performance test results can identify delays or problem areas.

Reviewer 2: ```suggestion
Performance test results help identify delays or areas of concern.
```

The plugin:warnings-ng[Warnings Next Generation] plugin provides convenient access to many reports, including:

* Compiler warnnings and errors (like gcc, clang, javac, or  golang)
* Static analysis warnings and errors (spotbugs, checkstyle, pmd, lint, cpd, or Simian)
* Code coverage reports

.How to use the Warnings Next Generation plugin
video::tj3xYFA6Q2o[youtube,width=800,height=420]

=== Build on agents

> Use agents to perform builds instead of running builds on the controller.

Reviewer 2: ```suggestion
It is recommended to use agents for performing builds instead of running them on the controller.
```

Author: I intentionally left this as more command form making an explicit statement of action, rather than a suggestion of possible action.

> It is much safer and more scalable to use agents.

Reviewer 2: ```suggestion
Utilizing agents offers enhanced safety and scalability.
```

More details are available in the link:/doc/book/security/controller-isolation/[controller isolation] documentation.

=== Show failures to the right people

> Configure the notifications from failing and unstable jobs so the notifications reach the right people without distracting others.

Reviewer 2: ```suggestion
Configure notifications for failing and unstable jobs to ensure that the right people receive them without causing unnecessary distractions for others.
```

> Many Jenkins users prefer to only be notified when a failure is likely their fault.

Reviewer 2: ```suggestion
 Many Jenkins users prefer to be notified only when a failure is likely their responsibility.
```

> They correctly assume that if they did not cause the failure, they are not the best person to investigate the failure.

Reviewer 2: ```suggestion
This approach acknowledges that if they are not responsible for the failure, they may not be the most suitable person to investigate it.
```

> Refine your notification system so that new test failures first notify the most recent committers, since they are the most likely cause of the issue.

Reviewer 2: ```suggestion
Refine your notification system to prioritize notifying the most recent committers when new test failures occur, as they are likely to be the cause of the issue.
```

.Sending Slack notifications
video::EDVZli8GdUM[youtube,width=800,height=420]

=== Use simple project names

> Jenkins uses project names for folders related to the project.

Reviewer 2: ```suggestion
Jenkins utilizes project names for organizing related folders.
```

> Many tools cannot handle spaces, dollar signs, or similar characters in file paths.

Reviewer 2: ```suggestion
However, it's important to note that certain tools may encounter issues with spaces, dollar signs, or similar characters in file paths.
```

> Limit yourself to alphanumeric characters [a-zA-Z0-9_-]+ in project names.

Reviewer 2: ```suggestion
To ensure compatibility, it's recommended to limit project names to alphanumeric characters` [a-zA-Z0-9_-]+`.
```

> Use the *Display Name* feature to make them look nice.

Reviewer 2: ```suggestion
To enhance the appearance of project names, you can utilize the *Display Name* feature.
This allows you to customize the presentation while maintaining the restricted characters in the underlying project name.
```

> The "Restrict project naming" setting in the system configuration can enforce naming restrictions for all projects.

Reviewer 2: ```suggestion
To enforce consistent naming conventions across all projects, you can enable the "Restrict project naming" setting in the system configuration. This will ensure that naming restrictions are enforced uniformly.
```

=== Fingerprint your dependencies

> When you have interdependent projects, it often becomes hard to keep track of which version of this is used by which version of that.

Reviewer 2: ```suggestion
When dealing with interdependent projects, it can be challenging to keep track of which version of one project is used by another.
```

> Jenkins supports "file fingerprinting" to simplify this.

Reviewer 2: ```suggestion
However, Jenkins offers a solution called "file fingerprinting" to simplify this process.
```

More information is available on the link:/doc/book/using/fingerprints/[fingerprinting page].

=== Don't use the Maven job type

> Jenkins has provided the plugin:maven-plugin[Maven integration plugin] for many years.
> Jobs created with the Maven integration plugin are created with the "Maven project" selection from the Jenkins "New item" menu.

Reviewer 2: ```suggestion
Jenkins has been providing the plugin:maven-plugin[Maven integration plugin] for many years, allowing users to create Maven projects using the "Maven project" selection from the Jenkins "New item" menu.
```

> The Maven job type provides a higher level of integration with maven builds, but creates unnecessary complications due to that higher level of integration.

Reviewer 2: ```suggestion
While the Maven job type offers a higher level of integration with Maven builds, it can sometimes introduce unnecessary complexities due to this deep integration.
```

> Use organization folders, multibranch Pipeline, or Pipeline jobs instead of the Maven job type.

Reviewer 2: ```suggestion
Consider using organization folders, multibranch Pipelines, or Pipeline jobs instead of the Maven job type. These alternatives provide more flexibility and simplicity in managing your Jenkins jobs and workflows.
```

The Jenkins project uses organization folders to build link:https://ci.jenkins.io/job/Core/[Jenkins core] and link:https://ci.jenkins.io/job/Plugins/[Jenkins plugins] on ci.jenkins.io.
A Jenkins Pipeline builds Maven projects easily and provides much better control for Maven users.

Refer to the link:https://plugins.jenkins.io/maven-plugin/#plugin-content-risks[Maven plugin documentation] for more details.

== Manage your controller

> The Jenkins controller is a central resource that needs to be managed for best results.

Reviewer 2: ```suggestion
The Jenkins controller plays a crucial role as a central resource and requires effective management for optimal performance.
```

> These practices can help your controller provide the best experience for users.

Reviewer 2: ```suggestion
By following these practices, you can ensure that your controller provides the best possible experience for users.
```

=== Secure the controller

> Jenkins installations have security enabled by default.
> However, it is possible to disable security.
> Do **not** disable security.

Reviewer 2: ```suggestion
Jenkins installations come with security enabled by default, which is a crucial aspect for protecting your system.
While it is technically possible to disable security, it is strongly advised **not** to do so.
Disabling security can leave your Jenkins instance vulnerable to unauthorized access and potential security breaches.
It is important to maintain a secure environment by keeping security enabled at all times.
```

Details are in the link:/doc/book/security/securing-jenkins/[securing Jenkins] chapter of the User Handbook.

=== Back up regularly

> Even the most reliable of systems will have failures.
> Prepare and regularly check the health of your backups.

Reviewer 2: ```suggestion
Even the most reliable systems can experience failures. 
That's why it's crucial to be prepared and regularly check the health of your backups.
Backups are a critical component of ensuring the integrity and availability of your data.
Regularly testing your backups and verifying their completeness and restorability will help you mitigate the impact of any potential failures and ensure that your data can be recovered effectively when needed.
Prioritizing backup health and conducting routine checks is essential for maintaining a robust and resilient system.
```

More details can be found in the link:/doc/book/system-administration/backing-up/[backup documentation].

=== Avoid scheduling overload

> Schedule jobs to balance the number of jobs running concurrently.

Reviewer 2: ```suggestion
Schedule your jobs strategically to balance the number of jobs running concurrently.
```

> If you're using timer triggers or are periodically polling, use the `H` syntax in the cron expression to introduce scheduing jitter.

Reviewer 2: ```suggestion
If you're using timer triggers or periodic polling, consider using the `H` syntax in the cron expression to introduce scheduling jitter.
This helps to distribute the start times of jobs more evenly and prevent them from all starting simultaneously.
```

> The predefined tokens such as `@hourly` can also be used to distribute job starting times more evenly.

Reviewer 2: ```suggestion
Additionally, take advantage of predefined tokens like `@hourly` to further distribute the starting times of your jobs.
These tokens can help create a more balanced schedule and reduce the likelihood of resource contention.

By implementing these scheduling techniques, you can optimize the utilization of your resources and ensure a smoother execution of your jobs.
```

=== Avoid resource collisions

> Multiple jobs running at the same time often cause collisions if they set up a service or need exclusive access.

Reviewer 2: ```suggestion
When multiple jobs run simultaneously, there is a possibility of collisions occurring, especially if they require exclusive access to certain resources or set-up services.
To prevent interference and ensure smooth execution, it is important to manage resource access effectively.
```

> If your builds involve the use of databases or other networked services, you need to ensure that they don't interfere with each other.

Reviewer 2: ```suggestion
For builds involving databases or networked services, it is crucial to implement measures that prevent conflicts.
```

> The plugin:lockable-resources[lockable resources plugin] provides fine grained resource locking for Jenkins jobs.

Reviewer 2: ```suggestion
The plugin:lockable-resources[Lockable Resources plugin] offers fine-grained resource-locking capabilities for Jenkins jobs.
By using this plugin, you can ensure that only one job has access to a specific resource at a time, avoiding conflicts and ensuring proper synchronization.
```

> If resource locking with the lockable resources plugin is not sufficient, you can prevent builds from running at the same time using the plugin:throttle-concurrent-builds[Throttle Concurrent Builds] plugin.

Reviewer 2: ```suggestion
In cases where resource locking with the lockable resources plugin is not sufficient, you can further control concurrent builds using the plugin:throttle-concurrent-builds[Throttle Concurrent Builds plugin].
This plugin allows you to limit the number of builds that can run simultaneously, providing additional control and preventing overload on shared resources.

By leveraging these plugins, you can manage resource conflicts and concurrency effectively, ensuring smooth and reliable execution of your Jenkins jobs.
```

Refer to the link:/doc/book/pipeline/pipeline-as-code/#organization-folders[organization folders] documentation for more details.

.GitHub organization folders
video::LbXKUKQ24T8[youtube,width=800,height=420]

.GitLab group folders
video::it6TOeQ6EHg[youtube,width=800,height=420]

.Bitbucket project folders
video::85b6fiVolfk[youtube,width=800,height=420]

.Gitea organization folders
video::NO3sZWRxgQM[youtube,width=800,height=420]

=== Use multibranch Pipelines

If you are unable to use organization folders, you can opt for multibranch Pipelines as an alternative.
However, it's important to note that organization folders are preferred over multibranch Pipelines because they provide the automation of creating and deleting multibranch projects when repositories are added or removed.

.GitHub multibranch Pipelines
video::aDmeeVDrp0o[youtube,width=800,height=420]

.GitLab group folders
video::y4XGFluzPHY[youtube,width=800,height=420]

.Bitbucket project folders
video::LNfthmZuRDI[youtube,width=800,height=420]

Refer to the link:/doc/book/pipeline/pipeline-as-code/#multibranch-pipeline-projects[multibranch Pipelines] documentation for more details.

=== Use Pipeline

> If organization folders are not an option for you, you can consider using multibranch Pipelines as an alternative.

Reviewer: ```suggestion
If organization folders are not an option for you, consider using multibranch Pipelines as an alternative.
```
small syntax suggestion

> However, it's important to highlight that organization folders are preferred due to their ability to automatically create and delete multibranch projects when repositories are added or removed.

Reviewer: ```suggestion
However, it's important to highlight that organization folders are preferred, due to their ability to automatically create and delete multibranch projects when repositories are added or removed.
```
suggestion for punctuation

Refer to the link:/doc/book/pipeline/[Pipeline] documentation for more details.

.Differences between Freestyle and Pipeline in Jenkins
video::IOUm1lw7F58[youtube,width=800,height=420]

== Manage your jobs

Jenkins job definitions can be managed and optimized to enhance user interactions and productivity.

=== Report build results

Charts and graphs provide valuable insights into project status and progress, showcasing trends and patterns.

> Automated test results, including unit tests, integration tests, and end-to-end tests, can reveal brittleness or instability. 

Reviewer: ```suggestion
Automated test results including unit tests, integration tests, and end-to-end tests can reveal brittleness or instability. 
```
suggestion for punctuation

Coverage reports help identify areas where automated tests are not being executed.
Compiler warning messages often serve as the first indication of a problem.
Static analysis tools are effective in reporting risky code or code with potential security risks.
Performance test results help identify delays or areas of concern.

> The plugin:warnings-ng[Warnings Next Generation] plugin provides convenient access to many reports, including:

Reviewer: ```suggestion
The plugin:warnings-ng[Warnings Next Generation] plugin provides convenient access to many reports including:
```
suggestion for punctuation

* Compiler warnnings and errors (like gcc, clang, javac, or  golang)
* Static analysis warnings and errors (spotbugs, checkstyle, pmd, lint, cpd, or Simian)
* Code coverage reports

.How to use the Warnings Next Generation plugin
video::tj3xYFA6Q2o[youtube,width=800,height=420]

=== Build on agents

Use agents to perform builds instead of running builds on the controller.
Utilizing agents offers enhanced safety and scalability.

> More details are available in the link:/doc/book/security/controller-isolation/[controller isolation] documentation.

Reviewer: ```suggestion
Refer to the link:/doc/book/security/controller-isolation/[controller isolation] documentation for more details.
```
suggestion for wording

=== Show failures to the right people

> Configure notifications for failing and unstable jobs to ensure that the right people receive them without causing unnecessary distractions for others.
>  Many Jenkins users prefer to be notified only when a failure is likely their responsibility.
> This approach acknowledges that if they are not responsible for the failure, they may not be the most suitable person to investigate it.

Reviewer: ```suggestion
Configure notifications for failing and unstable jobs, to ensure that the right people receive them without causing unnecessary distractions for others.
Many Jenkins users prefer to be notified only when a failure is likely their responsibility.
This approach acknowledges that if they are not responsible for the failure, they may not be the most suitable person to investigate it.
```
suggestion for punctuation/spacing

.Sending Slack notifications
video::EDVZli8GdUM[youtube,width=800,height=420]

=== Use simple project names

Jenkins utilizes project names for organizing related folders.
However, it's important to note that certain tools may encounter issues with spaces, dollar signs, or similar characters in file paths.
To ensure compatibility, it's recommended to limit project names to alphanumeric characters` [a-zA-Z0-9_-]+`.
To enhance the appearance of project names, you can utilize the *Display Name* feature.
This allows you to customize the presentation while maintaining the restricted characters in the underlying project name.
To enforce consistent naming conventions across all projects, you can enable the "Restrict project naming" setting in the system configuration. This will ensure that naming restrictions are enforced uniformly.

=== Fingerprint your dependencies

When dealing with interdependent projects, it can be challenging to keep track of which version of one project is used by another.
However, Jenkins offers a solution called "file fingerprinting" to simplify this process.

> More information is available on the link:/doc/book/using/fingerprints/[fingerprinting page].

Reviewer: ```suggestion
Refer to the link:/doc/book/using/fingerprints/[fingerprinting page] for more information.
```
suggestion for wording/sentence structure.

=== Don't use the Maven job type

Jenkins has been providing the plugin:maven-plugin[Maven integration plugin] for many years, allowing users to create Maven projects using the "Maven project" selection from the Jenkins "New item" menu.
While the Maven job type offers a higher level of integration with Maven builds, it can sometimes introduce unnecessary complexities due to this deep integration.

Consider using organization folders, multibranch Pipelines, or Pipeline jobs instead of the Maven job type. These alternatives provide more flexibility and simplicity in managing your Jenkins jobs and workflows.

The Jenkins project uses organization folders to build link:https://ci.jenkins.io/job/Core/[Jenkins core] and link:https://ci.jenkins.io/job/Plugins/[Jenkins plugins] on ci.jenkins.io.
A Jenkins Pipeline builds Maven projects easily and provides much better control for Maven users.

Refer to the link:https://plugins.jenkins.io/maven-plugin/#plugin-content-risks[Maven plugin documentation] for more details.

== Manage your controller

> The Jenkins controller plays a crucial role as a central resource and requires effective management for optimal performance.

Reviewer: ```suggestion
The Jenkins controller plays a crucial role as a central resource, requiring effective management for optimal performance.
```
suggestion for syntax

By following these practices, you can ensure that your controller provides the best possible experience for users.

=== Secure the controller

> Jenkins installations come with security enabled by default, which is a crucial aspect for protecting your system.

Reviewer: ```suggestion
Jenkins installations come with security enabled by default, which is a crucial aspect of protecting your system.
```
suggestion for word choice

While it is technically possible to disable security, it is strongly advised **not** to do so.
Disabling security can leave your Jenkins instance vulnerable to unauthorized access and potential security breaches.
It is important to maintain a secure environment by keeping security enabled at all times.

> Details are in the link:/doc/book/security/securing-jenkins/[securing Jenkins] chapter of the User Handbook.

Reviewer: ```suggestion
Refer to the link:/doc/book/security/securing-jenkins/[securing Jenkins] chapter of the User Handbook for more details.
```
suggestion for syntax

=== Back up regularly

Even the most reliable systems can experience failures. 
That's why it's crucial to be prepared and regularly check the health of your backups.
Backups are a critical component of ensuring the integrity and availability of your data.

> Regularly testing your backups and verifying their completeness and restorability will help you mitigate the impact of any potential failures and ensure that your data can be recovered effectively when needed.

Reviewer: ```suggestion
Regularly testing your backups and verifying their completeness and restorability, will help you mitigate the impact of any potential failures and ensure that your data can be recovered effectively when needed.
```
adding a comma since this is a longer sentence

Author: I disagree with the comma inserted here.

ifndef::env-github[:imagesdir: ../../resources]
endif::[]

Continuous Integration (CI) with automated test execution and trend analysis has revolutionized the way companies approach build management, release management, deployment automation, and test orchestration.
In this section, we will explore best practices that aim to enlighten executives, business managers, software developers, and architects about the invaluable contributions Jenkins can make throughout the project lifecycle.

== Automate job definition

Jenkins has the ability to automatically create, update, and delete jobs based on the repositories it identifies in your software configuration management system. 
Leverage this feature and optimize your job management by structuring your job definitions in a way that maximizes the benefits offered by Jenkins' automatic job management capabilities.

There are multiple alternatives for automatic job management, including:

* <<Use organization folders>> - create, update, and delete multibranch Pipeline folders and Pipeline jobs automatically (preferred)
* <<Use multibranch Pipelines>> - create, update and deleted Pipeline jobs automatically

> * <<Pipeline>> - Manually defined Pipeline jobs for more control over the job management process.

Author: Fix a hyperlink
```suggestion
* <<Use Pipeline>> - Manually defined Pipeline jobs for more control over the job management process.
```
